qiskit/circuit/singleton_gate.py

```diff
@@ -80,6 +80,25 @@ def __init__(self, *args, _condition=None, **kwargs):
         super().__init__(*args, **kwargs)
         self._condition = _condition
 
+    def __setstate__(self, state):
+        if state is not None:
+            self.label = state[0]
+            self._condition = state[1]
+            self.duration = state[2]
+            self.unit = state[3]
+
+    def __reduce__(self):
+        if not self.mutable:
+            return (self.__class__, (), None)
+        # Argument list is set twice to force instantiation via __setstate__
+        # the first arg list is only passed to __new__ to create a mutable
+        # object
+        return (
+            self.__class__,
+            (self.label, self._condition, self.duration, self.unit),
+            (self.label, self._condition, self.duration, self.unit),
+        )
```

Comment:
Did you have a particular reason to prefer `__setstate__`+`__reduce__` over `__getstate__`+`__setstate__`+`__getnewargs_ex__`?  The current form works fine so doesn't need changing, just asking because the docs suggest we should prefer the higher-level interface.

Reply:
Actually, I wasn't that familiar with the details of pickling, so I read the docs closely and I think there is a problem with the current form.

When this 3-tuple form is used for reduce, the second element of the tuple is unpacked and passed as the args to the first tuple element. Then `__setstate__` is called on the object returned by that call with the third tuple element passed in. The issue is that it is not guaranteed that all subclasses of `SingletonGate` take `(label, _condition, duration, unit)` as their first four arguments. For example, `test_mutable_pickle` uses `SXGate` which takes `(label, duration, unit, _condition)` (note the order is not the same as `SingletonGate`). When `pickle.load()` is called in the test, the second element of the reduce tuple gets passed to `SXGate` and triggers the mutable creation in `__new__` but then also gets passed to `SXGate.__init__` and initially creates an object with `_condition` and `label` exchanged.

`__getnewargs_ex__` could work if it were guaranteed that all subclasses would accept `(label, _condition, duration, unit)` as keyword arguments. For something more flexible, maybe `__reduce__` could return something like `(object.__new__, self.__class__, (self.label, self._condition, self.duration, self.unit))` so that unpickling still bypass `__init__` but creates a copy it can set the state on?

Reply:
Oh, that's true - I mistook that the first argument is the class object not specifically its `__new__`.  I think the reduction here would be fixed just by making the callable fetch `self.__class__.__new__` exactly to bypass `__init__`.  In practice, I don't think the current form will actually exhibit an issue, because the subsequent call to `__setstate__` will overwrite any mistakes during the `__init__`, but if the `__init__` started doing more type checking on its inputs, it could cause spurious exceptions before that stage.

I think the `__getstate__`/`__setstate__`/`__getnewargs_ex__` form might be a little easier - I think it's just the higher-level formalism of the system I described above with deliberately fetching `__class__.__new__`.

Reply:
I played with `__reduce__` return `(object.__new__, (self.__class__,), (self.label, self._condition, self.duration, self.unit))`. It kind of worked, but it helped me understand the point about pickle not calling `__init__`. `test_mutable_pickle` created a mutable `SXGate` and set the four attributes appropriately with `__setstate__`. However, the test then failed the instance did not have a `_name` attribute because only those four attributes were set as `super().__init__()` was never called to set the other attributes.

So now I understand why you advocate for the higher level methods. I think what would be perhaps most robust is defining `__getnewargs_ex__` to return a special keyword argument like `_create_mutable=True` (which hopefully no subclass would use) and then updating `__new__` to create a copy when this is passed (as well as the other conditions). Then we should not need `__getstate__` or `__setstate__` since the default implementation of copying attributes should be sufficient.

Reply:
I did it this way mostly because my knowledge of the pickle api is dated and I only knew how to do this with `__reduce__`. I wasn't aware of `__getnewargs_ex__`, but that interface is basically exactly what is needed here. I can refactor this to use the newer api.

Reply:
I updated this in: https://github.com/Qiskit/qiskit/pull/10871/commits/3ad38b4ae95f23bae69d47942162345ed71ca6dd we only need to define: `__getnewargs_ex__`  with that defined pickle will call `__new__` correctly so we either correctly use the singleton or the mutable copy returned. After that the default pickle mechanisms work fine as is.